text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,199 @@
+```rust
+async fn request_name(id: PersonID) -> String { .. }
+
+impl Requestable for async fn request_name {
+    /* ... */
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As described in the **Guide-level explanation**, with the syntax `[async] fn[(..) -> ?] <fn_path>`, we can reference the type behind the named function.
+
+When the function is for example in a different mod, it should be referenced by its path
+```rust
+mod sub {
+    fn sub_mod_fn() { .. }
+}
+trait MyTrait {}
+impl MyTrait for fn sub::sub_mod_fn {
+    /* ... */
+}
+```
+
+> ⚠️ NOTE: The same rules apply here as for normal types. Either the function item or the trait to implement mustn't be foreign for the impl. [Same as E0210](https://github.com/rust-lang/rust/blob/master/compiler/rustc_error_codes/src/error_codes/E0210.md)
+
+---
+
+It should be also possible to get the type of associated functions:
+
+```rust
+struct MyStruct;
+impl MyStruct {
+    fn new() -> Self { Self }
+}
+impl fn MyStruct::new {
+    /* ... */
+}
+```
+
+When the associated function comes from a trait, the same rules as for associated types apply here ([Ambiguous Associated Type, E0223](https://github.com/rust-lang/rust/blob/master/compiler/rustc_error_codes/src/error_codes/E0223.md)):
+
+```rust
+struct MyStruct;
+type MyTrait {
+    fn ambiguous();
+}
+impl MyTrait for MyStruct {
+    fn ambiguous() {  }
+}
+impl fn MyStruct::ambiguous { } // ERROR: ambiguous associated function
+// instead:
+impl fn <MyStruct as MyTrait>::ambiguous { } // OK
+```
+
+---
+
+When a function has generics, the function type is forced to be described, and the generic should be placed at it's desired position:
+```rust
+fn send<T: Send>(val: T, postal_code: u32) {}
+impl<T: Send> ParcelStation for fn(T, u32) send {
+    /* ... */
+}
+```
+
+When we have an implicit generic, the same rule applies
+```rust
+fn implicit_generic(val: impl Clone) -> impl ToString {}
+impl<T: Clone, U: ToString> for fn(T) -> U implicit_generic {
+    /* ... */
+}
+```
+
+---
+
+Just as structs and enums have the possibility to derive traits to automatically generate code, function type have similar ways via attribute macros:
+
+```rust
+#[debug_signature]
+fn signature_test(val: i32) -> bool {
+    /* ... */
+}
+
+// Expands to
+
+fn signature_test(val: i32) -> bool {
+    /* ... */
+}
+impl DbgSignature for fn signature_test {
+    fn dbg_signature() -> &'static str {
+        "fn signature_test(val: i32) -> bool"
+    }
+}
+```
+
+Other than that, it should behave like every other type does.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The type behind functions already exists, we just need to expose it to the user.
+
+# Prior art
+[prior-art]: #prior-art
+
+i dont know any
+
+# Unresolved questions
```

Comment:
I think this may need some work regarding lifetimes?

https://github.com/rust-lang/rust/issues/56105 is still unresolved, and I think it applies here as well?

Reply:
Ah yeah I'll add a lifetime example as well. Also I'll look if I need to link that tracking issue